File: 239.sliding-window-maximum.java

// @leet imports start
import java.util.*;
import java.math.*;
// @leet imports end

// @leet start

class Solution {
    private record Pair(int val, int idx) {
    }

    public int[] maxSlidingWindow(int[] nums, int k) {
        int n = nums.length;
        int[] ans = new int[n - k + 1];
        PriorityQueue<Pair> pq = new PriorityQueue<>((a, b) -> Integer.compare(b.val(), a.val()));

        for (int i = 0; i < n; i++) {
            Pair p = new Pair(nums[i], i);
            pq.add(p);

            if (i - k + 1 < 0) {
                continue;
            }

            while (!pq.isEmpty() && pq.peek().idx() < i - k + 1) {
                pq.poll();
            }

            ans[i - k + 1] = pq.peek().val();
        }

        return ans;

    }
}
// @leet end

File: 239.sliding-window-maximum.cpp
#include <vector>
using std::vector;
#include <deque>
using std::deque;

// @leet start
class Solution {
  public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        // O(n) TC from iteration
        // O(n) SC from deque
        // basically a monotonic queue where the front is always the largest

        deque<int> dq;
        vector<int> ans;
        auto n = nums.size();

        for (auto i = 0; i < n; ++i) {
            // remove all out of range elements in the front
            while (!dq.empty() && dq.front() <= i - k) {
                dq.pop_front();
            }
            // remove smaller numbers than current number
            // since curr number will be chosen rather than smaller numbers
            while (!dq.empty() && nums[dq.back()] < nums[i]) {
                dq.pop_back();
            }

            dq.push_back(i);

            if (i >= k - 1) {
                ans.push_back(nums[dq.front()]);
            }
        }

        return ans;
    }
};
// @leet end
